src/dialog.rs

```diff
@@ -0,0 +1,275 @@
+        let label = label.to_string();
+        let key = label.to_lowercase().chars().next().unwrap();
+        Self { label, key }
+    }
+    /// Create a new button with the given label and key.
+    pub fn with_key(label: &str, key: char) -> Self {
+        let label = label.to_string();
+        Self { label, key }
+    }
+}
+
+/// A dialog to display to the user
+///
+/// # Example
+/// ```rust
+/// use demand::Dialog;
+/// use demand::DialogButton;
+///
+/// let dialog = Dialog::new("Are you sure?")
+///   .description("This will do a thing.")
+///   .buttons(vec![
+///      DialogButton::new("Ok"),
+///      DialogButton::new("Not sure"),
+///      DialogButton::new("Cancel"),
+///   ]);
+/// let result = dialog.run().expect("error running confirm");
+/// ```
+pub struct Dialog<'a> {
+    /// The title of the selector
+    pub title: String,
+    /// The colors/style of the selector
+    pub theme: &'a Theme,
+    /// A description to display above the selector
+    pub description: String,
+    /// The buttons to display to the user
+    pub buttons: Vec<DialogButton>,
+
+    term: Term,
+    height: usize,
+    selected_button_idx: usize,
+}
+
+impl<'a> Dialog<'a> {
+    /// Create a new dialog with the given title
+    ///
+    /// By default, the dialog will have a single "Ok" button and a "Cancel" button.
+    pub fn new<S: Into<String>>(title: S) -> Self {
+        Self {
+            title: title.into(),
+            description: String::new(),
+            theme: &*theme::DEFAULT,
+            term: Term::stderr(),
+            buttons: vec![DialogButton::new("Ok"), DialogButton::new("Cancel")],
+            height: 0,
+            selected_button_idx: 0,
+        }
+    }
+
+    /// Set the description of the dialog
+    pub fn description(mut self, description: &str) -> Self {
+        self.description = description.to_string();
+        self
+    }
+
+    /// Set the buttons of the dialog
+    pub fn buttons(mut self, buttons: Vec<DialogButton>) -> Self {
+        self.buttons = buttons;
+        self
+    }
+
+    /// Set the index of the initially selected button.
+    ///
+    /// The `idx` is the index of the button in the `buttons` vector and is 0-indexed.
+    ///
+    /// # Errors
+    ///
+    /// This will panic if there are no buttons to select or if the index is out of bounds.
+    pub fn selected_button(mut self, idx: usize) -> Self {
+        if self.buttons.is_empty() {
+            panic!("No buttons to select");
+        }
+        if idx >= self.buttons.len() {
+            panic!("Selected button index out of bounds");
+        }
+        self.selected_button_idx = idx;
+        self
+    }
+
+    /// Set the theme of the dialog
+    pub fn theme(mut self, theme: &'a Theme) -> Self {
+        self.theme = theme;
+        self
+    }
+
+    /// Displays the dialog to the user and returns their response.
+    ///
+    /// The response will be the label of the selected button.
```

Comment:
Not sure if we need a value property on the DialogButton or if the label is good enough